packages/dd-trace/src/datastreams/schemas/schema_builder.js

```diff
@@ -0,0 +1,125 @@
+const LRUCache = require('lru-cache')
+const { fnv64 } = require('../fnv')
+const { Schema } = require('./schema')
+
+const maxDepth = 10
+const maxProperties = 1000
+const CACHE = new LRUCache({ max: 32 })
+
+class SchemaBuilder {
+  constructor (iterator) {
+    this.schema = new OpenApiSchema()
+    this.iterator = iterator
+    this.proerties = 0
+  }
+
+  addProperty (schemaName, fieldName, isArray, type, description, ref, format, enumValues) {
+    if (this.properties >= maxProperties) {
+      return false
+    }
+    this.properties += 1
+    let property = new OpenApiSchema.PROPERTY(type, description, ref, format, enumValues, null)
+    if (isArray) {
+      property = new OpenApiSchema.PROPERTY('array', null, null, null, null, property)
+    }
+    this.schema.components.schemas[schemaName].properties[fieldName] = property
+    return true
+  }
+
+  build () {
+    this.iterator.iterateOverSchema(this)
+    const noNones = convertToJsonCompatible(this.schema)
+    const definition = jsonStringify(noNones)
+    const id = fnv64(Buffer.from(definition, 'utf-8')).toString()
+    return new Schema(definition, id)
+  }
+
+  shouldExtractSchema (schemaName, depth) {
+    if (depth > maxDepth) {
+      return false
+    }
+    if (schemaName in this.schema.components.schemas) {
+      return false
+    }
+    this.schema.components.schemas[schemaName] = new OpenApiSchema.SCHEMA()
+    return true
+  }
+
+  static getSchema (schemaName, iterator) {
+    if (!CACHE.has(schemaName)) {
+      CACHE.set(schemaName, new SchemaBuilder(iterator).build())
+    }
+    return CACHE.get(schemaName)
+  }
+}
+
+class OpenApiSchema {
+  constructor () {
+    this.openapi = '3.0.0'
+    this.components = new OpenApiComponents()
+  }
+}
+
+OpenApiSchema.SCHEMA = class {
+  constructor () {
+    this.type = 'object'
+    this.properties = {}
+  }
+}
+
+OpenApiSchema.PROPERTY = class {
+  constructor (type, description = null, ref = null, format = null, enumValues = null, items = null) {
+    this.type = type
+    this.description = description
+    this.$ref = ref
+    this.format = format
+    this.enum = enumValues
+    this.items = items
+  }
+}
+
+class OpenApiComponents {
+  constructor () {
+    this.schemas = {}
+  }
+}
+
+function convertToJsonCompatible (obj) {
+  if (Array.isArray(obj)) {
+    return obj.filter(item => item !== null).map(item => convertToJsonCompatible(item))
+  } else if (obj && typeof obj === 'object') {
+    const jsonObj = {}
+    for (const [key, value] of Object.entries(obj)) {
+      if (value !== null) {
+        jsonObj[key] = convertToJsonCompatible(value)
+      }
+    }
+    return jsonObj
+  }
+  return obj
+}
+
+function convertKey (key) {
+  if (key === 'enumValues') {
+    return 'enum'
+  }
+  return key
+}
+
+function jsonStringify (obj, indent = 2) {
+  // made to stringify json exactly similar to python / java in order for hashing to be the same
```

Comment:
Oh, that's a good catch. Note that we can also remap things in the backend. But it's better if we are able to get the data right from the start.
Do you know if the order of fields will be the same between languages?

Reply:
No, its just based on the order the property was added. Should we sort them alphabetically?

Reply:
I think the current order works, it should be added in a fixed order for a given schema.